### ISSUE TITLE
[BUG] Custom link/interface attribute lost when using vlans in particular order

### ISSUE THREAD

**Issue by jbemmel** [ipspace/netlab#866]
### Describe the bug

In my use case I am trying to define a custom attribute 'epipe' on a link/interface. Given a certain combination of vlans on links, this attribute is lost in translation

### To Reproduce

```netlab create```

### Expected behavior

The 'epipe' attribute should show up in the topology file (ideally on the physical port like the 'mtu' link attribute, for some reason it gets copied to the vlan interface instead)

### Lab topology

```
message: |
  Illustrates an issue with a custom link attribute getting lost in translation

provider: clab

defaults:
  device: sros
  attributes.link.epipe: id

module: [vlan]

vlans:
  tenant:
    mode: bridge

groups:
  hosts:
    members: [ ce20 ]
  core:
    members: [ pe2 ]
  mtu:
    members: [ mtu1 ]

nodes:
 mtu1:
 pe2:
 ce20:

links:
- ce20: # Removing this one fixes the issue
  pe2:
  vlan.access: tenant
- mtu1:
  pe2:
   epipe: epipe-1     # This attribute is lost
   mtu: 1234
  vlan.access: tenant # This causes the epipe link attribute to get lost
```

### Output

In host_vars/pe2/topology.yml
```
hostname: clab-tmp-pe2
interfaces:
- bridge: tmp_1
  clab:
    name: eth1
  ifindex: 1
  ifname: 1/1/c1
  linkindex: 1
  mtu: 1500
  type: lan
  vlan:
    access: tenant
    access_id: 1000
- bridge: tmp_2
  clab:
    name: eth2
  ifindex: 2
  ifname: 1/1/c2
  linkindex: 2
  mtu: 1234
  type: lan
  vlan:
    access: tenant
    access_id: 1000
- bridge_group: 1
  ifindex: 3
  ifname: svi.1000
  name: VLAN tenant (1000) -> [ce20,mtu1]
  neighbors:
  - ifname: eth1
    ipv4: 172.16.0.3/24
    node: ce20
  - ifname: 1/1/c1
    ipv4: 172.16.0.1/24
    node: mtu1
  type: svi
  virtual_interface: true
  vlan:
    mode: bridge
```
Note how there is no 'epipe' attribute. Also see how the 'mtu' link attribute gets assigned to the lan interface, while the custom link attribute gets put on the vlan interface (when fixed, e.g. by removing the first link)

### Version

netlab version 1.6.1

### Additional context

I wasn't doing IBGP over EBGP this time...

**Comment by ipspace** [comment]
I think this is FAD (Functions as Designed) not a bug. It has to do with the way we create VLAN interfaces -- we copy attributes from a random physical interface that happens to be in that VLAN. Could you define that attribute on the VLAN (within a single node)?

**Comment by jbemmel** [comment]
We copy both interface-level and link-level attributes, and convert them into vlan interface attributes. Except for 'mtu', which is defined as a link-level attribute but gets applied to the physical interface when applied at the interface level

If I change the vlan on the ce20-pe2 link to a different one, it does copy the attributes.

It may be FAD, but it's hard to follow and somewhat inconsistent imho. In particular, I'd expect a custom interface attribute to be applied to the physical interface (like mtu), not the vlan

**Comment by jbemmel** [comment]
As an experiment, I rewrote ```create_svi_interfaces``` as follows:
```
def create_svi_interfaces(node: Box, topology: Box) -> dict:
  # skip_ifattr = list(topology.defaults.vlan.attributes.phy_ifattr)
  # skip_ifattr.extend(topology.defaults.providers.keys())

  copy_ifattr = [ "ipv4", "ipv6", "vlan_name", "gateway", "vrf" ] # removed 'pool'
  cleanup_ifattr = copy_ifattr + [ "neighbors", "name" ]

  vlan_ifmap: dict = {}

  # VLAN attributes not copied into VLAN interface: we take the global defaults and remove
  # mode attribute from that list as it's needed to set interface VLAN mode
  #
  svi_skipattr = [ k for k in list(topology.defaults.vlan.attributes.vlan_no_propagate or []) if k != "mode" ] + \
                  list(topology.defaults.vlan.attributes.vlan_svi_no_propagate)

  iflist_len = len(node.interfaces)
  for ifidx in range(0,iflist_len):
    ifdata = node.interfaces[ifidx]
    native_vlan = ifdata.get('vlan.native',None)
    access_vlan = ifdata.get('vlan.access',native_vlan)
    if not access_vlan:                                                     # No access VLAN on this interface?
      continue                                                              # ... good, move on

    routed_intf = interface_vlan_mode(ifdata,node,topology) == 'route'      # Identify routed VLAN interfaces
    vlan_subif  = routed_intf and ifdata.get('type','') == 'vlan_member'    # ... and VLAN-based subinterfaces

    vlan_data = create_node_vlan(node,access_vlan,topology)
    if vlan_data is None:                                                   # pragma: no-cover
      if vlan_subif:                                                        # We should never get here, but at least we can
        log.fatal(                                                       # scream before crashing
          f'Weird: cannot get VLAN data for VLAN {access_vlan} on node {node.name}, aborting')
      continue

    if routed_intf:                                                         # Routed VLAN access interface, turn it back into native interface
      for k in ('access','native'):                                         # ... remove access VLAN attributes
        ifdata.vlan.pop(k,None)

      if not ifdata.vlan:                                                   # ... and VLAN dictionary if there's nothing else left
        ifdata.pop('vlan',None)

      vlan_copy = get_box({ k:v for (k,v) in vlan_data.items() if not k in svi_skipattr and k != 'mode' })

      if vlan_subif:
        ifdata.vlan.mode = 'route'
        ifdata.vlan.access_id = vlan_data.id
      else:
        ifdata._vlan_mode = 'route'                                         # Flags we need to clean up up the routed native VLAN mess
        if native_vlan:                                                     # ... we can't use the vlan.* attributes as they might get removed
          ifdata._vlan_native = native_vlan

      node.interfaces[ifidx] = vlan_copy + ifdata                           # Merge VLAN data with interface data
      continue                                                              # Move to next interface

    features = devices.get_device_features(node,topology.defaults)
    svi_name = features.vlan.svi_interface_name
    if not svi_name:                                                        # pragma: no cover -- hope we got device settings right ;)
      log.error(                                                         # SVI interfaces are not supported on this device
        f'Device {node.device} used by {node.name} does not support VLAN interfaces (access vlan {access_vlan})',
        log.IncorrectValue,
        'vlan')
      return vlan_ifmap

    if access_vlan not in vlan_ifmap:                                       # Do we need to create a SVI interface?
      # skip_attr = list(skip_ifattr)                                         # Create a local copy of the attribute skip list
      vlan_mode = ifdata.vlan.get('mode','') or vlan_data.get('mode','')    # Get VLAN forwarding mode
      # if vlan_mode == 'bridge':                                             # ... and skip IP addresses for bridging-only VLANs
      #  skip_attr.extend(['ipv4','ipv6'])
        # continue  # JvB: in fact, skip creating SVI for L2-only VLANs
      vlan_ifdata = data.get_box(                                           # Copy non-physical interface attributes into SVI interface
        { k:ifdata[k] for k in copy_ifattr if k in ifdata })                # ... that will also give us IP addresses
      print( f"JvB: copied {vlan_ifdata}" )
      if vlan_mode:                                                         # Set VLAN forwarding mode for completness' sake
        vlan_ifdata.vlan.mode = vlan_mode
      vlan_ifdata.ifindex = node.interfaces[-1].ifindex + 1                 # Fill in the rest of interface data:
      vlan_ifdata.ifname = svi_name.format(                                 # ... ifindex, ifname, description
                              vlan=vlan_data.id,
                              bvi=vlan_data.bridge_group,
                              ifname=ifdata.ifname)
      vlan_ifdata.name = f'VLAN {access_vlan} ({vlan_data.id})'
      vlan_ifdata.virtual_interface = True                                  # Mark interface as virtual
      vlan_ifdata.type = "svi"
      vlan_ifdata.neighbors = []                                            # No neighbors so far
                                                                            # Overwrite interface settings with VLAN settings
      vlan_ifdata = vlan_ifdata + { k:v for k,v in vlan_data.items() if k not in svi_skipattr }
      fix_vlan_mode_attribute(vlan_ifdata)
      node.interfaces.append(vlan_ifdata)                                   # ... and add SVI interface to list of node interfaces
      vlan_ifmap[access_vlan] = vlan_ifdata
    else:
      vlan_ifdata = vlan_ifmap[access_vlan]

    update_vlan_neighbor_list(access_vlan,ifdata,vlan_ifdata,node,topology)

    for attr in cleanup_ifattr:                                             # Clean up physical interface data
      if attr in ifdata:
        print( f"JvB: cleanup {attr}" )
        ifdata.pop(attr,None)

    ifdata.vlan.access_id = vlan_data.id                                    # Add VLAN ID to interface data to simplify config templates

  return vlan_ifmap
```
i.e. instead of some elaborate logic around which attributes to skip, only define the small set to copy

Obviously we'd have to build in a way for the 'vrf' and 'gateway' module to add themselves to the list of things to copy, but other than that?

The tests pass (except for re-generating the ones that copy 'pool', unnecessarily I believe)

**Comment by ipspace** [comment]
> i.e. instead of some elaborate logic around which attributes to skip, only define the small set to copy

I totally agree with this idea. Going even further, we could (eventually) require most attributes be defined on the VLAN, not on a random interface. VLAN definition within a node would thus become very close to a VLAN/SVI interface.

Going a bit further, there might be L2 attributes that should be defined on the physical interface, everything else should be defined on the VLAN interface. However, MTU is not one of them (my bad) -- the MTU has to be consistent across whole L2 domain or someone is going to have a fun troubleshooting session.

Once we agree on that idea, we could start generating warnings about L3 attributes on VLAN access physical interfaces, then eventually turn that into errors and remove the copy logic. Doing that today might break stuff. Thoughts?

> Obviously we'd have to build in a way for the 'vrf' and 'gateway' module to add themselves to the list of things to copy, but other than that?

VRF is not a problem. It's a L3 attribute and should thus be defined on the VLAN.

Too early to think about gateways. In principle, gateway protocol is a L3 attribute, I just have to check what happens if you defined it on a VLAN.

> The tests pass (except for re-generating the ones that copy 'pool', unnecessarily I believe)

Pool is (again) a L3 attribute.

**Comment by jbemmel** [comment]
Re: pool, yes it's an L3 attribute but if it was already evaluated to generate the ip addresses, then it wouldn't need to be copied. If not, that means there is a missing test case ( address pools in combination with a VLAN ) - I'll look into that

I agree it's hard to see what would break today, even if all the tests pass

Re: vrf and gateway, my thinking was that for modularity reasons, it would be better if those modules defined how their attributes should be treated. However, there was already a precedent for 'gateway' processing logic in the vlan module - so adding 'vrf' isn't a big deviation

**Comment by ipspace** [comment]
> Re: pool, yes it's an L3 attribute but if it was already evaluated to generate the ip addresses, then it wouldn't need to be copied.

SVI interface generation is done after the links logic did its stuff, so the pool attribute is superfluous at that time, but it might be that someone is using it in a custom template for whatever other logic.

> If not, that means there is a missing test case ( address pools in combination with a VLAN ) - I'll look into that

No need I guess

> I agree it's hard to see what would break today, even if all the tests pass

Yeah, the test cases are what we consider somewhat-sane (or at least mainstream), who knows how people are using things.

> Re: vrf and gateway, my thinking was that for modularity reasons, it would be better if those modules defined how their attributes should be treated. However, there was already a precedent for 'gateway' processing logic in the vlan module - so adding 'vrf' isn't a big deviation

There is no need for extra VRF logic -- apply VRF attribute on the VLAN definition (like I usually do). Gateway is special because it's calculated for links and then has to be adjusted for VLANs (because I never wanted to implement the logic that would stitch together all VLAN segments into a single pseudo-link.

**Comment by ipspace** [comment]
On a totally tangential note: don't you think that the "epipe" attribute belongs to the VLAN, not to a VLAN access interface?

**Comment by jbemmel** [comment]
I’ve been struggling to come up with the right model, given attribute copying constraints. I may be able to mark a VLAN as being modeled as an epipe, and then define the evpn.eth_tag on the node (or interface, for convenience)This is for an EVPN VPWS serviceMaybe having those SVI interfaces for L2 services does make sense here after allSent from my iPhoneOn Aug 25, 2023, at 9:49 AM, Ivan Pepelnjak ***@***.***> wrote:﻿
On a totally tangential note: don't you think that the "epipe" attribute belongs to the VLAN, not to a VLAN access interface?

**Comment by ipspace** [comment]
> I’ve been struggling to come up with the right model, given attribute copying constraints. I thought about marking a VLAN as being modeled as an epipe, and then define the evpn.eth_tag on the node (or interface, for convenience)

That sounds about right.

> However, how would I suppress the output of the templates processing vlans? Those are unaware of the new 'epipe', and produce incompatible services (vpls in this case)

None of the templates I was involved with produce anything remotely similar to VPLS. They configure simple bridging, nothing more.

> This is for an EVPN VPWS service (epipe with bgp-evpn) stitched to a static VPWS (epipe) For now, I ended up defining a new type of global entities - "epipes" - very similar to (simplified) vlans, but not encumbered by the existing templates.

You could always change the VLAN template if you wish ;)

> An alternative would be to define generic top-level entities - say "services" - some of which would have type "vlan", others type "epipe", etc. - then we could have an extensible service model, in which plugins append a new allowed type string for service.type

Not saying NO, but this needs to be thought out pretty carefully or we'll end in another VLAN mess ;)

**Comment by ipspace** [comment]
Made sure only the mandatory attributes (addresses, first-hop gateway) are copied from PHY to VLAN interface. Also made the "don't do it" warning in the documentation more explicit.

Generating an error message when encountering parameters on PHY interface would be too cumbersome because a lot of parameters get copied from VLAN definition to access links in early parts of the transformation process.
